index.js

```diff
@@ -0,0 +1,111 @@
+import express from "express";
+import cors from "cors";
+import books from "./books.js";
+import path from "path";
+
+const app = express();
+const PORT = 8080;
+
+// parse JSON
+app.use(express.json());
+// parse URL encoded data
+app.use(express.urlencoded({ extended: true }));
+
+app.use(cors());
+
+// to let the server know wh
+const __dirname = path.resolve();
+app.use(express.static("client"));
+
+// api/books
+// create an endpoint the /api/book endpoint - Get request
+
+app.get("/api/books", (req, res) => {
+  res.json(books);
+});
+
+app.get("/", (req, res) => {
+  res.sendFile(path.join(__dirname, "client", "index.html"));
+});
+
+// Adding Books
+app.post("/book", (req, res) => {
+  const id = books.length + 1;
+  const newBook = req.body;
+  newBook.id = id;
+  // console.log(newBook);
+  books.push(newBook);
+  return res.redirect("/");
+});
+
+// Retrieving a Book by id
+app.get("/book/:id", (req, res) => {
+  try {
+    const id = parseInt(req.params.id);
+    //  console.log(id);
+    let book = books.find((book) => book.id === id);
+    console.log(book);
+    if (!book) {
+      return res.status(404).json({
+        message: "Book not found",
+      });
+    }
+
+    return res.json(book);
+  } catch (error) {
+    res.status(500).json({
+      message: "Failed to retrieve book",
+    });
+  }
+});
+
+// Editing Books
+app.post("/book/:id", (req, res) => {
+  try {
+    const id = parseInt(req.params.id);
+    console.log(id);
+    let book = books.find((book) => book.id === id);
+    console.log(book);
+    if (!book) {
+      return res.status(404).json({
+        message: "Book not found",
+      });
+    }
+    const bookIDX = books.indexOf(book);
+    books[bookIDX].title = req.body.title || books[bookIDX].title;
+    books[bookIDX].author = req.body.author || books[bookIDX].author;
+    books[bookIDX].format = req.body.format || books[bookIDX].format;
+    books[bookIDX].isbn = req.body.isbn || books[bookIDX].isbn;
+
+    res.redirect("/");
```

Comment:
This API is mixing return types which would cause problems in real-world usage. When there's an error, it returns JSON, but when it's successful, it returns a redirect header. Not a problem for this exercise, but just wanted to point it out.